rust/src/import_parsing.rs

```diff
@@ -0,0 +1,595 @@
+use crate::errors::GrimpResult;
+use ruff_python_ast::statement_visitor::{walk_body, walk_stmt, StatementVisitor};
+use ruff_python_ast::{Expr, Stmt};
+use ruff_python_parser::parse_module;
+use ruff_source_file::{LineIndex, SourceCode};
+use std::fs;
+use std::path::Path;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct ImportedObject {
+    pub name: String,
+    pub line_number: usize,
+    pub line_contents: String,
+    pub typechecking_only: bool,
+}
+
+impl ImportedObject {
+    fn new(
+        name: String,
+        line_number: usize,
+        line_contents: String,
+        typechecking_only: bool,
+    ) -> Self {
+        Self {
+            name,
+            line_number,
+            line_contents,
+            typechecking_only,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ImportedObjectWithoutLineContents {
+    pub name: String,
+    pub line_number: usize,
+    pub typechecking_only: bool,
+}
+
+impl ImportedObjectWithoutLineContents {
+    fn new(name: String, line_number: usize, typechecking_only: bool) -> Self {
+        Self {
+            name,
+            line_number,
+            typechecking_only,
+        }
+    }
+}
+
+pub fn parse_imports(path: &Path) -> GrimpResult<Vec<ImportedObject>> {
+    let code = fs::read_to_string(path).expect("failed to read file");
+    parse_imports_from_code(&code)
+}
+
+pub fn parse_imports_from_code(code: &str) -> GrimpResult<Vec<ImportedObject>> {
+    let imports_without_line_contents = parse_imports_from_code_without_line_contents(code)?;
+
+    let lines: Vec<&str> = code.lines().collect();
+
+    Ok(imports_without_line_contents
+        .into_iter()
+        .map(|i| {
+            ImportedObject::new(
+                i.name,
+                i.line_number,
+                lines[i.line_number - 1].trim_start().to_string(),
+                i.typechecking_only,
+            )
+        })
+        .collect())
+}
+
+fn parse_imports_from_code_without_line_contents(
+    code: &str,
+) -> GrimpResult<Vec<ImportedObjectWithoutLineContents>> {
+    let ast = parse_module(code).expect("failed to parse python code");
```

Comment:
Do you know if the Python syntax it checks is the same as the version of Python it will be running under?

Reply:
Good question. I'm unsure, but I'd suspect the latest python syntax is parsed. Assuming python doesn't release a new major version, then the syntax should always be backwards compatible? So it should be fine I think.